lib/eco/pgviewer.py

```diff
@@ -0,0 +1,135 @@
+from threading import Thread
+
+class PGThread(Thread):
+
+    def __init__(self, gl):
+        Thread.__init__(self)
+        self.gl = gl
+
+    def run(self):
+        self.gl.display()
+
+    def quit(self):
+        drawgraph.display_async_quit()
+
+class PGViewer(object):
+
+    def __init__(self, root, version):
+        self.root = root
+
+        self.whitespaces = True
+        self.version = version
+        self.restrict_nodes = []
+        self.ast = None
+        self.links = {}
+        self.gl = drawgraph.GraphLayout(0,0,0)
+        self.gl.pgviewer = self
+        self.max_version = 0
+        self.t = None
+
+    def quit(self):
+        self.t.quit()
+
+    def run(self):
+        self.create_graph()
+        self.t = PGThread(self.gl)
+        self.t.start()
+
+    def is_alive(self):
+        if not self.t:
+            return False
+        return self.t.is_alive()
+
+    def send(self, command, *args):
+        if command == "refresh":
+            self.refresh(self.version + args[0])
+        elif command == "toggle_ws":
+            self.whitespaces = self.whitespaces ^ True
+            self.refresh(self.version)
+
+    def refresh(self, version):
+        if self.is_alive():
+            if version > 0 and version <= self.max_version:
+                self.version = version
+                self.create_graph()
+                self.gl.gdisplay.zoom(self.gl.scale)
+                self.gl.gdisplay.updated_viewer()
+                self.gl.gdisplay.redraw_now()
+
+    def create_graph(self):
+        self.links = {}
+        self.gl.reset()
+        graph = pydot.Dot(graph_type='digraph')
+        self.add_node(self.root, graph)
+
+        lines = graph.create_dot("plain").splitlines(True)
+
+        _, scale, width, height = lines[0].strip("\n").split(" ")
+        self.gl.scale = float(scale)
+        self.gl.boundingbox = float(width), float(height)
+        self.gl.links = self.links
+
+        for line in lines[1:]:
+            args = splitline(line)
+            if args[0] == "node":
+                self.gl.add_node(*args[1:])
+            elif args[0] == "edge":
+                self.gl.add_edge(*args[1:])
+
+    def add_node(self, node, graph):
+        self.links["%s" % id(node)] = node
+        try:
+            label = node.get_attr("symbol.name", self.version)
+            lookup = node.get_attr("lookup", self.version)
+        except Exception:
+            label = node.symbol.name
+            lookup = node.lookup
+        if label == "\r":
+            label = "<return>"
+        if lookup == "<ws>":
+            label = "<ws> x%s" % len(label)
+        label = label.replace("\"", "\\\"")
+        label = label.replace("\\\\\"", "\\\\\\\"")
+        label = label.replace("\r", "\\n")
+        dotnode = pydot.Node("\"%s\"" % id(node), label='"%s"' % label, shape="box")
+
+        try:
+            changed = node.get_attr("changed", self.version)
+        except Exception:
```

Comment:
catching an ignoring every kind of exception here hides errors, just catch a precise type.